### PR TITLE
chore(build): generate api artifacts

### LINKED ISSUE #63: Remove committed generated code — generate proto and OpenAPI at build time
## User Request

Delete the OpenAPI schema and all generated code (protobuf) from the repo. Proto should be generated from buf at build time. OpenAPI should be pulled from prebuilt OCI charts in `agynio/api` at build time.

**Architecture reference:** `agynio/architecture` → `architecture/api-contracts.md`
> "Services **do not** commit generated schema code in their own repositories. They consume schemas as dependencies at build time."

---

### 1. Overview

The repository currently commits three categories of generated artifacts:

| Category | Committed path | Generator | Upstream source |
|----------|---------------|-----------|-----------------|
| Protobuf Go stubs | `gen/` | `buf generate` | `buf.build/agynio/api` |
| OpenAPI YAML specs (embedded) | `internal/apischema/` | `scripts/sync-embedded-spec.sh` | `ghcr.io/agynio/openapi/{team,llm}:1` |
| oapi-codegen server stubs | `internal/gen/server.gen.go`, `internal/llmgen/server.gen.go` | `oapi-codegen` | `.openapi/*.yaml` |

After this change, **none** of these artifacts will be committed. All three will be generated at build time in every context: Docker build, CI, and DevSpace.

---

### 2. Files and directories to delete

| # | Path | Type | Reason |
|---|------|------|--------|
| 1 | `gen/` | Directory (entire tree) | Committed protobuf stubs; already in `.gitignore`, already generated at build time. |
| 2 | `internal/apischema/` | Directory (entire tree) | Embedded OpenAPI YAML copies and Go `spec.go` loader files. Replaced by build-time generation. |
| 3 | `internal/gen/server.gen.go` | File | oapi-codegen generated Team server code. Will be generated at build time. |
| 4 | `internal/llmgen/` | Directory (entire tree) | oapi-codegen generated LLM server code. Will be generated at build time. |
| 5 | `oapi-codegen.server.yaml` | File | oapi-codegen config. Replaced by CLI flags in generation script. |
| 6 | `oapi-codegen.llm.yaml` | File | oapi-codegen config. Same as above. |
| 7 | `scripts/sync-embedded-spec.sh` | File | Copies pulled specs into `internal/apischema/`. No longer needed. |
| 8 | `.spectral.yaml` | File | Spectral lint config. Linting now happens in `agynio/api`. |

### Files to preserve (NOT delete)

| File | Reason |
|------|--------|
| `internal/gen/extra_types.go` | Hand-written type aliases required by oapi-codegen output. |
| `internal/gen/workspace_limit_json_test.go` | Hand-written test exercising generated types. |

---

### 3. New build-time generation script

Create `scripts/generate-openapi.sh`:

```bash
#!/usr/bin/env bash
# generate-openapi.sh — Pull OpenAPI specs from GHCR and generate Go server code.
# Requires: oras, oapi-codegen, gofmt on $PATH.
set -euo pipefail

SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
ROOT_DIR="$(cd "${SCRIPT_DIR}/.." && pwd)"

# ── Step 1: Pull specs ──────────────────────────────────────────────
bash "${SCRIPT_DIR}/pull-spec.sh"

# ── Step 2: Place specs for //go:embed ──────────────────────────────
mkdir -p "${ROOT_DIR}/internal/apischema/teamv1"
mkdir -p "${ROOT_DIR}/internal/apischema/llmv1"
cp "${ROOT_DIR}/.openapi/team-v1.yaml" "${ROOT_DIR}/internal/apischema/teamv1/team-v1.yaml"
cp "${ROOT_DIR}/.openapi/llm-v1.yaml"  "${ROOT_DIR}/internal/apischema/llmv1/llm-v1.yaml"

# ── Step 3: Generate spec.go loaders ────────────────────────────────
cat > "${ROOT_DIR}/internal/apischema/teamv1/spec.go" << 'GOEOF'
package teamv1

import (
	_ "embed"

	"github.com/getkin/kin-openapi/openapi3"
)

//go:embed team-v1.yaml
var Spec []byte

func LoadSpec() (*openapi3.T, error) {
	loader := openapi3.NewLoader()
	return loader.LoadFromData(Spec)
}
GOEOF

cat > "${ROOT_DIR}/internal/apischema/llmv1/spec.go" << 'GOEOF'
package llmv1

import (
	_ "embed"

	"github.com/getkin/kin-openapi/openapi3"
)

//go:embed llm-v1.yaml
var Spec []byte

func LoadSpec() (*openapi3.T, error) {
	loader := openapi3.NewLoader()
	return loader.LoadFromData(Spec)
}
GOEOF

# ── Step 4: Run oapi-codegen ────────────────────────────────────────
mkdir -p "${ROOT_DIR}/internal/llmgen"

oapi-codegen \
  -package gen \
  -generate chi-server,strict-server,models \
  -o "${ROOT_DIR}/internal/gen/server.gen.go" \
  "${ROOT_DIR}/.openapi/team-v1.yaml"

oapi-codegen \
  -package llmgen \
  -generate chi-server,strict-server,models \
  -o "${ROOT_DIR}/internal/llmgen/server.gen.go" \
  "${ROOT_DIR}/.openapi/llm-v1.yaml"

# ── Step 5: Format generated code ──────────────────────────────────
gofmt -w "${ROOT_DIR}/internal/gen/server.gen.go"
gofmt -w "${ROOT_DIR}/internal/llmgen/server.gen.go"

echo "OpenAPI generation complete."
```

---

#### 4.1 `Dockerfile`

Add ORAS and oapi-codegen installation. Call `scripts/generate-openapi.sh` before the Go build:

```dockerfile
# syntax=docker/dockerfile:1.7
ARG GO_VERSION=1.22
ARG BUF_VERSION=1.64.0
ARG ORAS_VERSION=1.1.0
ARG OAPI_CODEGEN_VERSION=v2.4.0

# Stage 1: Download buf binary
FROM --platform=$BUILDPLATFORM golang:${GO_VERSION}-bookworm AS buf
ARG BUF_VERSION
RUN curl -sSL \
      "https://github.com/bufbuild/buf/releases/download/v${BUF_VERSION}/buf-$(uname -s)-$(uname -m)" \
      -o /usr/local/bin/buf && \
    chmod +x /usr/local/bin/buf

# Stage 2: Generate + compile
FROM --platform=$BUILDPLATFORM golang:${GO_VERSION}-bookworm AS build
WORKDIR /src
ENV CGO_ENABLED=0 GO111MODULE=on
ARG ORAS_VERSION
ARG OAPI_CODEGEN_VERSION

COPY --from=buf /usr/local/bin/buf /usr/local/bin/buf

# Install ORAS
RUN curl -fsSL -o /tmp/oras.tar.gz \
      "https://github.com/oras-project/oras/releases/download/v${ORAS_VERSION}/oras_${ORAS_VERSION}_linux_$(dpkg --print-architecture).tar.gz" && \
    tar -xzf /tmp/oras.tar.gz -C /usr/local/bin oras && \
    rm /tmp/oras.tar.gz

# Install oapi-codegen
RUN go install github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen@${OAPI_CODEGEN_VERSION}

COPY go.mod go.sum ./
RUN --mount=type=cache,target=/root/.cache/go-build \
    --mount=type=cache,target=/root/go/pkg/mod \
    go mod download && go mod verify

# Copy only the files needed for generation first
COPY buf.gen.yaml buf.yaml ./
COPY scripts/ scripts/

# Generate protobuf stubs
RUN buf generate buf.build/agynio/api \
      --path agynio/api/files/v1 \
      --path agynio/api/llm/v1 \
      --path agynio/api/secrets/v1 \
      --path agynio/api/teams/v1

# Generate OpenAPI code (pull specs + oapi-codegen + embed wrappers)
RUN bash scripts/generate-openapi.sh

# Copy remaining source (hand-written files only; generated files already in place)
COPY . .

ARG TARGETOS TARGETARCH
RUN --mount=type=cache,target=/root/.cache/go-build \
    --mount=type=cache,target=/root/go/pkg/mod \
    GOOS=$TARGETOS GOARCH=$TARGETARCH go build \
      -trimpath -ldflags="-s -w" -o /out/gateway ./cmd/gateway

# Stage 3: Runtime
FROM gcr.io/distroless/static-debian12:nonroot
WORKDIR /app
LABEL org.opencontainers.image.source="https://github.com/agynio/gateway"
COPY --from=build /out/gateway ./gateway
EXPOSE 8080
ENTRYPOINT ["/app/gateway"]
```

#### 4.2 `.github/workflows/ci.yml`

1. Remove "Verify generated code" steps (no committed code to diff against)
2. Remove "Verify embedded spec matches GHCR" step
3. Remove "Spectral lint" step (linting happens in `agynio/api`)
4. Remove `@stoplight/spectral-cli` from npm install
5. Remove oasdiff breaking-change steps (owned by `agynio/api`)
6. Replace "Pull OpenAPI spec from GHCR" with call to `scripts/generate-openapi.sh`

Updated step order:
```
1. Checkout
2. Set up Go
3. Set up Node.js (for Redoc build only)
4. Install toolchain (oapi-codegen, redocly)
5. Install ORAS
6. Set up Buf
7. Generate gRPC stubs (buf generate — unchanged)
8. Generate OpenAPI code (oras login + scripts/generate-openapi.sh)
9. Go vet
10. Go test
11. Build Redoc documentation
12. Upload Redoc artifact
```

#### 4.3 `scripts/devspace-startup.sh`

Add OpenAPI generation:
```bash
#!/usr/bin/env bash
set -eu

echo "=== DevSpace startup ==="

echo "Generating protobuf types..."
buf generate buf.build/agynio/api \
  --path agynio/api/files/v1 \
  --path agynio/api/llm/v1 \
  --path agynio/api/secrets/v1 \
  --path agynio/api/teams/v1

echo "Generating OpenAPI code..."
bash scripts/generate-openapi.sh

echo "Downloading Go modules..."
go mod download

echo "Starting dev server (air)..."
exec air
```

#### 4.4 `devspace.yaml`

Add generated paths to exclude lists (`internal/apischema/`, `internal/gen/server.gen.go`, `internal/llmgen/`).

#### 4.5 `.gitignore`

```gitignore
.venv/
.hypothesis/
dist/
.openapi/
gateway.log
gateway.pid

# Build-time generated code — do not commit
gen/
internal/apischema/
internal/gen/server.gen.go
internal/llmgen/
```

Note: Do NOT ignore entire `internal/gen/` — it contains hand-written `extra_types.go` and `workspace_limit_json_test.go`.

#### 4.6 `.dockerignore`

Add generated paths to prevent locally-generated files leaking into build context:
```
internal/apischema/
internal/gen/server.gen.go
internal/llmgen/server.gen.go
```

#### 4.7 `buf.gen.yaml`

Remove the comment `# proto-update: teams-v1-variables-2026-03-13`.

#### 4.8 `README.md`

Update development workflow documentation to reflect build-time generation.

#### 4.9 `cmd/gateway/main.go`

**No changes needed.** Import paths remain the same — generated code lands at the same paths.

---

### 5. Verification checklist

| # | Check |
|---|-------|
| 1 | No generated files in git: `git ls-files gen/ internal/apischema/ internal/gen/server.gen.go internal/llmgen/` returns only `internal/gen/extra_types.go` and `internal/gen/workspace_limit_json_test.go` |
| 2 | `go vet ./...` passes after generation |
| 3 | `go test ./...` passes (specifically `internal/gen/` workspace limit tests) |
| 4 | CI workflow passes |

---

### 6. Edge cases

- **`extra_types.go`**: Declares `package gen`, same as generated `server.gen.go`. No conflict.
- **`workspace_limit_json_test.go`**: References generated types — only compiles after generation. Fine for CI/Docker.
- **`oapi-codegen` version pinning**: Must remain `v2.4.0` across Dockerfile and CI.
- **Dockerfile `COPY . .` ordering**: Runs after generation. `.dockerignore` prevents local generated files from overwriting.
- **`internal/llmgen/` empty in git**: Created by `mkdir -p` in script. No `.gitkeep` needed.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- remove committed generated proto/OpenAPI artifacts
- add build-time OpenAPI generation script and wire Docker/CI/devspace flows
- update ignores and README for build-time generation

## Testing
- nix shell nixpkgs#buf -c buf generate buf.build/agynio/api --path agynio/api/files/v1 --path agynio/api/llm/v1 --path agynio/api/secrets/v1 --path agynio/api/teams/v1
- nix shell nixpkgs#go nixpkgs#gcc nixpkgs#oras -c bash -c 'go install github.com/oapi-codegen/oapi-codegen/v2/cmd/oapi-codegen@v2.4.0; export PATH="$(go env GOPATH)/bin:$PATH"; bash scripts/generate-openapi.sh'
- nix shell nixpkgs#go nixpkgs#gcc -c go vet ./...
- nix shell nixpkgs#go nixpkgs#gcc -c go test ./...

Closes #63

### CODE REVIEW

File: scripts/generate-openapi.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+//go:embed llm-v1.yaml
+var Spec []byte
+
+func LoadSpec() (*openapi3.T, error) {
+	loader := openapi3.NewLoader()
+	return loader.LoadFromData(Spec)
+}
+GOEOF
+
+# -- Step 4: Run oapi-codegen -----------------------------------------
```
Comment (noa-lucent) [inline scripts/generate-openapi.sh]: [major] Missing `mkdir -p` for `internal/gen/` before the `oapi-codegen` call.

During the Docker build, `internal/gen/` doesn't exist at this point — only `go.mod`, `go.sum`, `buf.gen.yaml`, `buf.yaml`, and `scripts/` have been `COPY`ed. The hand-written files (`extra_types.go`, `workspace_limit_json_test.go`) that normally keep this directory alive aren't available until `COPY . .`, which runs *after* generation.

`internal/llmgen` gets a `mkdir -p` on line 57, but `internal/gen` does not. This will cause a "no such file or directory" error in the Docker build.

Add before the first `oapi-codegen` call:
```bash
mkdir -p "${ROOT_DIR}/internal/gen"
```

**Comment by casey-brooks** [comment]
## Summary
- ensure scripts/generate-openapi.sh creates internal/gen before codegen

## Test & Lint Summary
- Lint: nix shell nixpkgs#go nixpkgs#gcc -c go vet ./... (no errors)
- Tests: nix shell nixpkgs#go nixpkgs#gcc -c go test ./... (passed: 3, failed: 0, skipped: 0)

File: scripts/generate-openapi.sh (hunk 1 above)
Comment (rowan-stein) [reply scripts/generate-openapi.sh]: Fixed in bebd7ee — added `mkdir -p "${ROOT_DIR}/internal/gen"` before the first `oapi-codegen` call.

**Review by noa-lucent — DISMISSED** [review]
Fix verified — `mkdir -p` for `internal/gen/` is now in place alongside `internal/llmgen/`. All prior feedback resolved. LGTM.

**Comment by casey-brooks** [comment]
## End-to-End Test Results
- Bootstrap: `bootstrap_v2/apply.sh -y` completed successfully; kubeconfig merged.
- Cluster health: `kubectl get pods -n platform` shows gateway running; `kubectl get application gateway -n argocd` is Synced/Healthy.
- DevSpace: `devspace dev` completed (startup installed oras + oapi-codegen, generated OpenAPI, gateway listening on :8080).
- Gateway check: `kubectl port-forward -n platform svc/gateway-gateway 8080:8080` + `curl http://localhost:8080/team/v1/agents` returned JSON payload.

## Test & Lint Summary
- Lint: `nix shell nixpkgs#go nixpkgs#gcc -c go vet ./...` (no errors)
- Tests: `nix shell nixpkgs#go nixpkgs#gcc -c go test ./...` (passed: 3, failed: 0, skipped: 0)

File: scripts/devspace-startup.sh [hunk 2]
```diff
@@ -3,9 +3,26 @@ set -eu
 
 echo "=== DevSpace startup ==="
 
+export PATH="$(go env GOPATH)/bin:$PATH"
+ORAS_VERSION="1.1.0"
+OAPI_CODEGEN_VERSION="v2.4.0"
```
Comment (noa-lucent) [inline scripts/devspace-startup.sh]: [minor] These version constants are now duplicated in three places (Dockerfile, CI, and here). When a version is bumped, all three must be updated in sync — easy to miss.

Consider extracting them into a single source of truth, e.g. a `.env` or `scripts/versions.env` file that all three contexts source/reference. Not blocking, but worth a follow-up to reduce drift risk.

**Review by noa-lucent — APPROVED** [review]
New commit looks good. The guard-and-install pattern for `oras` and `oapi-codegen` is appropriate for a dev container startup script, version pins match across all three contexts, and the `go install` module paths are correct.

One minor note on version duplication across files — see inline comment. Not blocking.
